File: engine/src/mcts.cpp
#include "mcts.h"
#include <algorithm>
#include <cmath>
#include <random>

namespace chess {

Search::Search(const MCTSConfig& cfg, std::vector<Position>&& game_history,
               std::mt19937_64* rng)
    : cfg_(cfg), hist_(std::make_unique<std::vector<Position>>(std::move(game_history))),
      rng_(rng) {
    root_ = tree_.new_node();
}

uint32_t Search::node_visits(const Tree::Node& n, const Tree::Edge* es) {
    uint32_t v = 0;
    for (int i = 0; i < n.num; ++i) v += es[i].visits;
    return v;
}

float Search::cpuct(float parent_visits) const {
    return cfg_.cpuct_init +
           std::log((parent_visits + cfg_.cpuct_base + 1.0f) / cfg_.cpuct_base);
}

int Search::root_visits() const {
    return static_cast<int>(node_visits(tree_.node(root_), tree_.edges(root_)));
}

float Search::root_q() const {
    const Tree::Node& n = tree_.node(root_);
    const Tree::Edge* es = tree_.edges(root_);
    double s = 0;
    uint32_t v = 0;
    for (int i = 0; i < n.num; ++i) { s += es[i].value_sum; v += es[i].visits; }
    return v ? static_cast<float>(s / v) : 0.0f;
}

int Search::select_child(int node, std::vector<PathEntry>& path) {
    const Tree::Node& n = tree_.node(node);
    Tree::Edge* es = tree_.edges(node);

    uint32_t total_v = 0;
    double total_s = 0.0;
    for (int i = 0; i < n.num; ++i) { total_v += es[i].visits; total_s += es[i].value_sum; }

    float parent_q = total_v ? static_cast<float>(total_s / total_v) : 0.0f;
    float fpu = parent_q - cfg_.fpu_reduction;
    float sqrt_n = std::sqrt(static_cast<float>(std::max<uint32_t>(total_v, 1u)));
    float c = cpuct(static_cast<float>(total_v));

    int best = -1;
    float best_score = -1e30f;
    for (int i = 0; i < n.num; ++i) {
        const Tree::Edge& e = es[i];
        float q = e.visits ? e.value_sum / static_cast<float>(e.visits) : fpu;
        float u = c * (static_cast<float>(e.prior) * (1.0f / 65535.0f)) * sqrt_n /
                  (1.0f + static_cast<float>(e.visits));
        float score = q + u;
        if (score > best_score) { best_score = score; best = i; }
    }
    path.push_back({node, best});
    return best;
}

void Search::backprop(const std::vector<PathEntry>& path, float value_from_leaf_stm) {
    float contrib = -value_from_leaf_stm;
    for (auto it = path.rbegin(); it != path.rend(); ++it) {
        Tree::Edge& e = tree_.edges(it->node)[it->edge];
        e.visits += 1;
        e.value_sum += contrib + cfg_.virtual_loss;
        contrib = -contrib;
    }
}

EvalTask Search::make_task(const Position& leaf, const std::vector<Position>& walk,
                           const std::vector<Move>& legal, int node_idx,
                           const std::vector<PathEntry>& path) {
    EvalTask t;
    t.moves.reserve(legal.size());
    t.priors.assign(legal.size(), 0.0f);
    for (const Move& m : legal) t.moves.push_back(pack_move(m));
    t.stm = leaf.stm;
    t.root_stm = hist_->back().stm;
    t.path.reserve(path.size());
    for (const PathEntry& pe : path) t.path.emplace_back(pe.node, pe.edge);

    t.enc = encode_position(leaf, *hist_, walk);
    t.node = node_idx;
    tree_.node(node_idx).state = Tree::ST_INFLIGHT;
    return t;
}

void Search::compensate_vl(const std::vector<PathEntry>& path) {
    for (auto it = path.rbegin(); it != path.rend(); ++it) {
        Tree::Edge& e = tree_.edges(it->node)[it->edge];
        e.value_sum += cfg_.virtual_loss;
    }
}

void Search::descend_once(std::vector<EvalTask>& out_tasks) {
    thread_local std::vector<PathEntry> path;
    thread_local std::vector<Position> walk;
    path.clear();
    walk.clear();
    int node = root_;

    auto cur = [&]() -> const Position& { return walk.empty() ? hist_->back() : walk.back(); };

    for (;;) {
        Tree::Node& n = tree_.node(node);

        if (n.state == Tree::ST_TERMINAL) {
            backprop(path, n.term_value);
            return;
        }
        if (n.state == Tree::ST_INFLIGHT) {
            compensate_vl(path);
            return;
        }
        if (n.state == Tree::ST_UNEXPANDED) {
            const Position& leaf = cur();
            thread_local std::vector<Move> legal;
            legal.clear();
            gen_legal_moves(leaf, legal);
            if (legal.empty()) {
                bool check = attacks::square_attacked(
                    leaf, leaf.king_sq(static_cast<Color>(leaf.stm)),
                    static_cast<Color>(leaf.stm ^ 1), leaf.occ());
                n.state = Tree::ST_TERMINAL;
                n.term_value = check ? -1.0f : draw_value();
                backprop(path, n.term_value);
                return;
            }
            out_tasks.push_back(make_task(leaf, walk, legal, node, path));
            return;
        }

        int ei = select_child(node, path);
        Tree::Edge& e = tree_.edges(node)[ei];
        e.value_sum -= cfg_.virtual_loss;

        if (e.child < 0) e.child = tree_.new_node();

        Position next = make_move(cur(), unpack_move(e.move));

        if (cfg_.use_twofold_draw) {
            int seen = 0;
            for (const Position& p : *hist_)
                if (p.hash == next.hash) ++seen;
            for (const Position& p : walk)
                if (p.hash == next.hash) ++seen;
            if (seen >= 1) {   // was >= 2 (threefold); this is a search-internal twofold prune
                Tree::Node& cn = tree_.node(e.child);
                cn.state = Tree::ST_TERMINAL;
                cn.term_value = draw_value_for(next.stm);
                backprop(path, cn.term_value);
                return;
            }
        }
        if (next.halfmove >= 100) {
            Tree::Node& cn = tree_.node(e.child);
            cn.state = Tree::ST_TERMINAL;
            cn.term_value = draw_value_for(next.stm);
            backprop(path, cn.term_value);
            return;
        }

        walk.push_back(std::move(next));
        node = e.child;
    }
}

void Search::gather_round(int max_leaves, std::vector<EvalTask>& out_tasks) {
    for (int i = 0; i < max_leaves; ++i)
        descend_once(out_tasks);
}

void Search::maybe_noise_root() {
    Tree::Node& n = tree_.node(root_);
    if (n.noised || !cfg_.root_dirichlet) return;
    if (n.num <= 0) return;

    std::gamma_distribution<float> gamma(cfg_.dirichlet_alpha, 1.0f);
    std::vector<float> noise(n.num);
    float sum = 0;
    for (auto& x : noise) { x = gamma(*rng_); sum += x; }
    if (sum <= 0) return;
    for (auto& x : noise) x /= sum;

    Tree::Edge* es = tree_.edges(root_);
    for (int i = 0; i < n.num; ++i) {
        float p = (1.0f - cfg_.dirichlet_eps) * (es[i].prior / 65535.0f) +
                  cfg_.dirichlet_eps * noise[i];
        es[i].prior = static_cast<uint16_t>(
            std::min(65535.0f, std::max(0.0f, p * 65535.0f)));
    }
    n.noised = 1;
}

void Search::complete_task(EvalTask&& t) {
    Tree::Node& n = tree_.node(t.node);
    if (n.state != Tree::ST_INFLIGHT) return;

    std::vector<Tree::Edge> es;
    es.reserve(t.moves.size());
    for (size_t i = 0; i < t.moves.size(); ++i) {
        Tree::Edge e;
        e.move = t.moves[i];
        Move m = unpack_move(t.moves[i]);
        e.policy_idx = static_cast<int16_t>(move_policy_index(m, static_cast<Color>(t.stm)));
        e.prior = static_cast<uint16_t>(
            std::min(65535.0f, std::max(0.0f, t.priors[i] * 65535.0f)));
        e.child = -1;
        es.push_back(e);
    }
    tree_.attach_edges(t.node, std::move(es));
    n.state = Tree::ST_EXPANDED;
    n.task_slot = -1;

    if (t.node == root_) maybe_noise_root();

    std::vector<PathEntry> path;
    path.reserve(t.path.size());
    for (const auto& [nd, ed] : t.path) path.push_back({nd, ed});
    backprop(path, t.value);
}

Search::Choice Search::pick_move(std::mt19937_64& rng, int ply_in_game) {
    const Tree::Node& n = tree_.node(root_);
    const Tree::Edge* es = tree_.edges(root_);
    Choice ch;
    if (n.num <= 0) return ch;

    std::vector<uint32_t> visits(n.num);
    uint32_t total = 0;
    for (int i = 0; i < n.num; ++i) { visits[i] = es[i].visits; total += visits[i]; }

    int pick = -1;
    if (ply_in_game < cfg_.prior_plies) {
        uint32_t ptot = 0;
        for (int i = 0; i < n.num; ++i) ptot += es[i].prior;
        if (ptot > 0) {
            std::uniform_int_distribution<uint32_t> dist(1, ptot);
            uint32_t roll = dist(rng);
            for (int i = 0; i < n.num; ++i) {
                if (roll <= es[i].prior) { pick = i; break; }
                roll -= es[i].prior;
            }
            if (pick < 0) pick = n.num - 1;
        }
    }

    if (pick < 0 && ply_in_game < cfg_.temperature_plies && total > 0) {
        std::uniform_int_distribution<uint32_t> dist(1, total);
        uint32_t roll = dist(rng);
        for (int i = 0; i < n.num; ++i) {
            if (roll <= visits[i]) { pick = i; break; }
            roll -= visits[i];
        }
        if (pick < 0) pick = n.num - 1;
    }

    if (pick < 0) {
        uint32_t bestv = 0;
        for (int i = 0; i < n.num; ++i)
            if (visits[i] > bestv) { bestv = visits[i]; pick = i; }
        if (pick < 0) pick = 0;
    }

    ch.move = es[pick].move;
    ch.chosen_edge = pick;
    ch.child_node = es[pick].child;
    return ch;
}

void Search::apply_root_move(uint16_t packed_move, int child_node) {
    Position np = make_move(hist_->back(), unpack_move(packed_move));
    hist_->push_back(std::move(np));

    bool reuse = (child_node >= 0 && tree_.node(child_node).state == Tree::ST_EXPANDED);
    if (reuse) {
        tree_.keep_subtree(child_node);
        root_ = 0;   // keep_subtree always relabels the retained root to index 0
        maybe_noise_root();
    } else {
        tree_.clear();
        root_ = tree_.new_node();
    }
}

void Search::root_visit_distribution(std::vector<std::pair<int, int>>& out) const {
    out.clear();
    const Tree::Node& n = tree_.node(root_);
    const Tree::Edge* es = tree_.edges(root_);
    out.reserve(n.num);
    for (int i = 0; i < n.num; ++i)
        if (es[i].policy_idx >= 0 && es[i].visits > 0)
            out.emplace_back(es[i].policy_idx, static_cast<int>(es[i].visits));
}

void Tree::keep_subtree(int new_root) {
    std::vector<Node> nn;
    std::vector<Edge> ne;
    nn.reserve(nodes_.size() / 4 + 64);
    ne.reserve(edges_.size() / 4 + 64);

    std::vector<int> old2new(nodes_.size(), -1);
    std::vector<int> stack;
    old2new[new_root] = 0;
    nn.push_back(nodes_[new_root]);
    stack.push_back(new_root);

    while (!stack.empty()) {
        int o = stack.back();
        stack.pop_back();
        int mapped = old2new[o];
        Node cp = nodes_[o];
        if (cp.num > 0) {
            const Edge* old_es = edges_.data() + cp.first;
            int32_t first_new = static_cast<int32_t>(ne.size());
            for (int i = 0; i < cp.num; ++i) {
                Edge e = old_es[i];
                if (e.child >= 0) {
                    if (old2new[e.child] < 0) {
                        old2new[e.child] = static_cast<int>(nn.size());
                        nn.push_back(nodes_[e.child]);
                        stack.push_back(e.child);
                    }
                    e.child = old2new[e.child];
                }
                ne.push_back(e);
            }
            cp.first = first_new;
        }
        nn[mapped] = cp;
    }
    nodes_.swap(nn);
    edges_.swap(ne);
}

}  // namespace chess

File: serve/mcts.py
"""Python port of engine/src/mcts.cpp for serving a trained checkpoint.

This mirrors the C++ `Search` class closely enough to be a faithful port:
  - PUCT selection with the same cpuct/FPU formulas (mcts.cpp Search::select_child)
  - virtual loss during batched leaf gathering (mcts.cpp Search::backprop /
    compensate_vl), so a single-threaded search can still gather a full batch
    of diverse leaves per round before calling the network
  - search-internal twofold-repetition + 50-move pruning
    (mcts.cpp Search::descend_once)

Deliberately DOES NOT port:
  - root Dirichlet noise (self-play exploration only)
  - prior-ply / temperature move sampling (self-play exploration only)
  - tree reuse across plies (`keep_subtree`) - each HTTP request is a fresh
    search from scratch, since the frontend is stateless per move

Move selection here is always "greedy": the root edge with the most visits
(ties broken by prior), i.e. what engine/src/selfplay.cpp's `pick_move` does
once past all the exploration windows.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Optional

import chess

from trainer.encoding import _pos_key, move_policy_index

ST_UNEXPANDED, ST_INFLIGHT, ST_EXPANDED, ST_TERMINAL = 0, 1, 2, 3


class Edge:
    __slots__ = ("move", "prior", "policy_idx", "visits", "value_sum", "child")

    def __init__(self, move: chess.Move, prior: float, policy_idx: int):
        self.move = move
        self.prior = prior
        self.policy_idx = policy_idx
        self.visits = 0
        self.value_sum = 0.0
        self.child: Optional[int] = None


class Node:
    __slots__ = ("state", "term_value", "edges")

    def __init__(self):
        self.state = ST_UNEXPANDED
        self.term_value = 0.0
        self.edges: list[Edge] = []


@dataclass
class EvalTask:
    node: int
    path: list[tuple[int, int]]
    board: chess.Board             # leaf position (board.turn == stm)
    legal: list[chess.Move]
    policy_idx: list[int]
    stm: bool                      # chess.WHITE / chess.BLACK
    root_stm: bool
    history: list[chess.Board]     # game history + search walk, ends with `board`


class Search:
    CPUCT_INIT = 1.25
    CPUCT_BASE = 19652.0
    FPU_REDUCTION = 0.2
    VIRTUAL_LOSS = 1.0

    def __init__(self, root_board: chess.Board, game_history: list[chess.Board],
                 use_twofold_draw: bool = True):
        """`game_history` must end with a board equal to `root_board` (same
        convention as the engine's `hist_`: full game so far, root position last)."""
        self.root_board = root_board.copy(stack=False)
        self.history = game_history
        self.use_twofold_draw = use_twofold_draw
        self.nodes: list[Node] = [Node()]
        self.root = 0

    # -- small helpers ---------------------------------------------------
    def _cpuct(self, parent_visits: float) -> float:
        return self.CPUCT_INIT + math.log(
            (parent_visits + self.CPUCT_BASE + 1.0) / self.CPUCT_BASE)

    def _draw_value_for(self, stm: bool, contempt: float) -> float:
        # Search-side draw value with contempt, mirroring the blend used for
        # real WDL predictions in onnx_model.priors_and_value (a rule-based
        # draw is just a pure-draw WDL distribution run through the same
        # formula). contempt=0.5 (the default here) is neutral / no bias.
        c_eff = contempt if stm == self.root_board.turn else (1.0 - contempt)
        return 2.0 * c_eff - 1.0

    def _new_node(self) -> int:
        self.nodes.append(Node())
        return len(self.nodes) - 1

    @staticmethod
    def _repeats(boards: list[chess.Board], board: chess.Board) -> int:
        key = _pos_key(board)
        return sum(1 for b in boards if _pos_key(b) == key)

    def root_visits(self) -> int:
        return sum(e.visits for e in self.nodes[self.root].edges)

    def root_q(self) -> float:
        """Average backed-up value from the root's side-to-move perspective."""
        edges = self.nodes[self.root].edges
        v = sum(e.visits for e in edges)
        if v == 0:
            return 0.0
        s = sum(e.value_sum for e in edges)
        return s / v

    def pick_move_greedy(self) -> Edge:
        edges = self.nodes[self.root].edges
        return max(edges, key=lambda e: (e.visits, e.prior))

    # -- PUCT selection ----------------------------------------------------
    def _select_child(self, node_idx: int, path: list[tuple[int, int]]) -> int:
        n = self.nodes[node_idx]
        edges = n.edges
        total_v = sum(e.visits for e in edges)
        total_s = sum(e.value_sum for e in edges)
        parent_q = (total_s / total_v) if total_v else 0.0
        fpu = parent_q - self.FPU_REDUCTION
        sqrt_n = math.sqrt(max(total_v, 1))
        c = self._cpuct(total_v)

        best, best_score = -1, -1e30
        for i, e in enumerate(edges):
            q = (e.value_sum / e.visits) if e.visits else fpu
            u = c * e.prior * sqrt_n / (1.0 + e.visits)
            score = q + u
            if score > best_score:
                best_score, best = score, i
        path.append((node_idx, best))
        return best

    def _backprop(self, path: list[tuple[int, int]], value_from_leaf_stm: float) -> None:
        contrib = -value_from_leaf_stm
        for node_idx, ei in reversed(path):
            e = self.nodes[node_idx].edges[ei]
            e.visits += 1
            e.value_sum += contrib + self.VIRTUAL_LOSS
            contrib = -contrib

    def _compensate_vl(self, path: list[tuple[int, int]]) -> None:
        for node_idx, ei in reversed(path):
            self.nodes[node_idx].edges[ei].value_sum += self.VIRTUAL_LOSS

    # -- one simulation ------------------------------------------------------
    def descend_once(self, contempt: float) -> Optional[EvalTask]:
        path: list[tuple[int, int]] = []
        walk: list[chess.Board] = []
        node = self.root

        def cur() -> chess.Board:
            return walk[-1] if walk else self.root_board

        while True:
            n = self.nodes[node]

            if n.state == ST_TERMINAL:
                self._backprop(path, n.term_value)
                return None
            if n.state == ST_INFLIGHT:
                self._compensate_vl(path)
                return None
            if n.state == ST_UNEXPANDED:
                board = cur()
                legal = list(board.legal_moves)
                if not legal:
                    term_value = -1.0 if board.is_check() else self._draw_value_for(board.turn, contempt)
                    n.state, n.term_value = ST_TERMINAL, term_value
                    self._backprop(path, term_value)
                    return None
                n.state = ST_INFLIGHT
                policy_idx = [move_policy_index(mv, board) for mv in legal]
                return EvalTask(
                    node=node, path=list(path), board=board, legal=legal,
                    policy_idx=policy_idx, stm=board.turn, root_stm=self.root_board.turn,
                    history=self.history + walk)

            ei = self._select_child(node, path)
            edge = n.edges[ei]
            edge.value_sum -= self.VIRTUAL_LOSS
            if edge.child is None:
                edge.child = self._new_node()

            nb = cur().copy(stack=False)
            nb.push(edge.move)

            if self.use_twofold_draw:
                # "seen before" set = full game history (incl. root) + this
                # descent's walk so far, NOT including `nb` itself.
                seen = self._repeats(self.history + walk, nb)
                if seen >= 1:
                    cn = self.nodes[edge.child]
                    cn.state = ST_TERMINAL
                    cn.term_value = self._draw_value_for(nb.turn, contempt)
                    self._backprop(path, cn.term_value)
                    return None
            if nb.halfmove_clock >= 100:
                cn = self.nodes[edge.child]
                cn.state = ST_TERMINAL
                cn.term_value = self._draw_value_for(nb.turn, contempt)
                self._backprop(path, cn.term_value)
                return None

            walk.append(nb)
            node = edge.child

    def gather_round(self, max_leaves: int, contempt: float) -> list[EvalTask]:
        out: list[EvalTask] = []
        for _ in range(max_leaves):
            t = self.descend_once(contempt)
            if t is not None:
                out.append(t)
        return out

    def complete_task(self, task: EvalTask, priors: list[float], value: float) -> None:
        n = self.nodes[task.node]
        if n.state != ST_INFLIGHT:
            return
        n.edges = [Edge(mv, p, pidx)
                   for mv, p, pidx in zip(task.legal, priors, task.policy_idx)]
        n.state = ST_EXPANDED
        self._backprop(task.path, value)